File: src/sketchup-stl/SKUI/js/bridge.js
/*******************************************************************************
 *
 * module Bridge
 *
 ******************************************************************************/


var Bridge = function() {

  /* Due to OSX' async callback nature the messages needs to be queued up and
   * carefully sent in sequence. The next message cannot be sent until a message
   * from Ruby confirms the last message was received.
   */
  var messages = [];
  var busy = false;

  return {


    /* Prepares the communication bridge.
     */
    init : function() {
      bridge = $('<input id="SKUI_RUBY_BRIDGE" type="hidden" />');
      $('body').append( bridge );
    },


    /* Escapes backslashes and single quotes.
     * (private ?)
     */
    escape_string : function( value ) {
      return value.replace(/\\/g, '\\\\').replace(/'/g, "\\'");
    },


    /* Executes a Javascript command and returns the return value to the
     * Ruby bridge.
     */
    execute : function( code_string ) {
      // Clean up <SCRIPT></SCRIPT> elements which Ruby
      // UI::WebDialog.execute_script leaves behind.
      $('body script').detach();
      // Execute the JavaScript code and put the return value back into the
      // bridge.
      Bridge.return_ruby( eval(code_string) );
    },


    /* Returns the checked state for the Checkbox control given by UI ID.
     */
    get_checkbox_state : function( ui_id ) {
      return $('#'+ui_id+' input').prop('checked');
    },


    /* Returns the checked state for the given jQuery selector.
     */
    get_checked_state : function( selector ) {
      return $(selector).prop('checked');
    },


    /* Returns the HTML for the given jQuery selector.
     */
    get_html : function( selector ) {
      return $(selector).html();
    },


    /* Returns the Rect for the Checkbox control given by UI ID.
     */
    get_control_rect : function( ui_id ) {
      $control = $('#'+ui_id);
      position = $control.position();
      width  = $control.outerWidth();
      height = $control.outerHeight();
      rect = {
        'left'   : position.left,
        'top'    : position.top,
        'right'  : position.left + width,
        'bottom' : position.top + height,
        'width'  : width,
        'height' : height
      }
      return rect;
    },


    /* Returns the text for the given jQuery selector.
     */
    get_text : function( selector ) {
      return $(selector).text();
    },


    /* Returns the value for the given jQuery selector.
     */
    get_value : function( selector ) {
      return $(selector).val();
    },


    /* Process the next message in the queue to Ruby.
     */
    push_message : function() {
      if ( busy ) {
        return false;
      } else {
        message = messages.shift();
        if ( message ) {
          busy = true;
          uri_message = encodeURIComponent( message.add_slashes() );
          window.location = 'skp:SKUI_Callback@' + uri_message;
          return true;
        } else {
          return false;
        }
      }
    },


    /* Ruby calls this when a message has been received which means the next
     * message can be sent.
     */
    pump_message : function() {
      busy = false;
      Bridge.push_message();
      return busy;
    },


    /* Send the next message in the queue to Ruby.
     */
    queue_message : function( message ) {
      messages.push( message );
      return Bridge.push_message();
    },


    /* Resets the Ruby bridge.
     */
    reset : function() {
      $('#SKUI_RUBY_BRIDGE').val( '' );
    },


    /* Returns a Javascript object to the Ruby bridge element so that SketchUp
     * Ruby script can fetch the value.
     */
    return_ruby : function( value ) {
      $('#SKUI_RUBY_BRIDGE').val( Bridge.value_to_ruby(value) );
    },


    /* Sets the ID for the <BODY> element so that the Window class' properties
     * can control the appearance of the window background.
     */
    set_window_id : function( ui_id ) {
      $('body').attr( 'id', ui_id );
    },


    /* Converts Javascript objects into Ruby objects.
     */
    value_to_ruby : function( value ) {
      var ruby_string = '';
      switch ( $.type( value ) ) {
        case 'boolean':
          ruby_string = value.toString();
          break;
        case 'number':
          if ( isNaN( value ) ) {
            ruby_string = '0.0/0.0';
          } else if ( isFinite( value ) ) {
            ruby_string = value.toString();
          } else {
            // Infinite
            ruby_string = ( value > 0 ) ? '1.0/0.0' : '-1.0/0.0';
          }
          break;
        case 'string':
          ruby_string = "'" + Bridge.escape_string( value ) + "'";
          break;
        case 'null':
        case 'undefined':
          ruby_string = 'nil';
          break;
        case 'array':
          ruby_values = $.map(value, function(array_value, index) {
            return Bridge.value_to_ruby( array_value );
          });
          ruby_string = '[' + ruby_values.join(',') + ']';
          break;
        case 'date':
          ruby_string = 'Time.at(' + value.getTime() + ')';
          break;
        case 'regexp':
          // http://www.w3schools.com/jsref/jsref_obj_regexp.asp
          var i = value.ignoreCase ? 'i' : '';
          var g = value.global     ? 'g' : ''; // Not supported in Ruby.
          var m = value.multiline  ? 'm' : '';
          var regex = '/'+value.source+'/'+i+m;
          ruby_string = Bridge.value_to_ruby( regex );
          break;
        case 'function':
          ruby_string = "'<FUNCTION>'";
          break;
        case 'object':
          // Assume JSON.
          var ruby_key = '';
          var ruby_value = '';
          var ruby_hash = new Array();
          $.each(value, function(k, v) {
            ruby_key   = Bridge.value_to_ruby( k );
            ruby_value = Bridge.value_to_ruby( v );
            ruby_hash.push( ruby_key + " => " + ruby_value );
          });
          ruby_string = '{' + ruby_hash.join(',') + '}';
          break;
      }
      return ruby_string;
    }


  };

}(); // Bridge

File: src/sketchup-stl/SKUI/bridge.rb
module SKUI

  require File.join( PATH, 'control.rb' )
  require File.join( PATH, 'control_manager.rb' )
  require File.join( PATH, 'json.rb' )

  
  # Handles the communication between Ruby and the WebDialog.
  #
  # @since 1.0.0
  class Bridge

    # @since 1.0.0
    class CommunicationError < StandardError; end

    # @since 1.0.0
    attr_accessor( :webdialog, :window )

    # @param [SKUI::Window] window
    # @param [UI::WebDialog] webdialog
    #
    # @since 1.0.0
    def initialize( window, webdialog )
      @window = window
      @webdialog = webdialog
    end

    # Wrapper to build a script string and return the return value of the called
    # Javascript function.
    #
    # This method also ensures a that the +<SCRIPT>+ elements which
    # +UI::WebDialog.execute_script+ leaves behind is cleaned up.
    #
    #  return_value = window.bridge.call('alert', 'Hello World')
    #
    # @param [String] function Name of JavaScript function to call.
    # @param [Mixed] args List of arguments for the function call.
    #
    # @return [Mixed]
    # @since 1.0.0
    def call( function, *args )
      # A JavaScript command is prepared and sent to the JS bridge which then
      # evaluates it and puts the return value into a hidden <input> element
      # which is then pulled back from Ruby and evaluated as Ruby objects.

      # Ensure that we don't pull old data back from the WebDialog in case there
      # is stale data due to some previous error.
      @webdialog.execute_script( 'Bridge.reset()' )
      # (!) SU-0415
      # Reports of .execute_script might have a hard limit - possibly under OSX
      # only. Windows does seem unaffected. 
      # Test case:
      #  w.execute_script("alert('#{'x'*10000000}'.length);")
      arguments = args.map { |arg| JSON.object_to_js( arg ) }.join(',')
      javascript = "#{function}(#{arguments});".inspect
      # If WebDialog is not visible, or no HTML is populated (lacking DOM) then
      # .execute_script returns false.
      #
      # (i) OSX - SU6
      # http://forums.sketchucation.com/viewtopic.php?f=180&t=8316#p49259
      # Indicates that ; might cause the call to fail. Seems to work without,
      # so keeping it like that to be on the safe size.
      # puts "Bridge.execute(#{javascript})" #DEBUG
      unless @webdialog.execute_script( "Bridge.execute(#{javascript})" )
        if @webdialog.visible?
          raise( CommunicationError, 'Window not visible.' )
        else
          raise( CommunicationError, 'Unknown error. Ensure DOM is ready.' )
        end
      end
      # (?) Catch JavaScript errors? Or just let the WebDialog display the error?
      raw_data = @webdialog.get_element_value( 'SKUI_RUBY_BRIDGE' )
      @webdialog.execute_script( 'Bridge.reset()' )
      # The JS Bridge converts the JS values into Ruby code strings.
      # (?) Catch exceptions? Re-raise with custom exception?
      eval( raw_data ) # (?) Bind to top level scope?
    end

    # @param [Control] control
    #
    # @return [Nil]
    # @since 1.0.0
    def add_control( control )
      call( 'UI.add_control', control.properties )
      nil
    end

    # @param [ControlManager] container
    #
    # @return [Nil]
    # @since 1.0.0
    def add_container( container )
      # (?) Compile into one large function call, might it be faster to execute?
      for control in container.controls
        add_control( control )
        if control.is_a?( ControlManager )
          add_container( control )
        end
      end
      nil
    end

    # @param [String] selector jQuery selector
    #
    # @return [String] Returns the HTML code for the given jQuery selector.
    # @since 1.0.0
    def get_html( selector )
      call( 'Bridge.get_html', selector )
    end

    # @param [String] ui_id ID to a +Checkbox+ control.
    #
    # @return [String] Returns the checked state for the given Control.ui_id.
    # @since 1.0.0
    def get_checkbox_state( ui_id )
      call( 'Bridge.get_checkbox_state', ui_id )
    end

    # @param [String] selector jQuery selector
    #
    # @return [String] Returns the checked state for the given jQuery selector.
    # @since 1.0.0
    def get_checked_state( selector )
      call( 'Bridge.get_checked_state', selector )
    end
    
    # @param [String] selector jQuery selector
    #
    # @return [String] Returns the text content for the given jQuery selector.
    # @since 1.0.0
    def get_text( selector )
      call( 'Bridge.get_text', selector )
    end

    # It appear that under OSX UI::WebDialog.get_element_value doesn't work for
    # <TEXTAREA> and <SELECT> elements. Using this instead solves the issue.
    #
    # @param [String] selector jQuery selector
    #
    # @return [String] Returns the value for the given jQuery selector.
    # @since 1.0.0
    def get_value( selector )
      call( 'Bridge.get_value', selector )
    end

    # @param [String] ui_id Control.ui_id
    #
    # @return [Hash] Returns the Rect for the given Control.
    # @since 1.0.0
    def get_control_rect( ui_id )
      call( 'Bridge.get_control_rect', ui_id )
    end

    # @param [String] ui_id Control.ui_id
    #
    # @return [String] Returns the value for the given Control.
    # @since 1.0.0
    def get_control_value( ui_id )
      get_value( "##{ui_id}" )
    end
    alias :get_element_value :get_control_value

    # @return [Nil]
    # @since 1.0.0
    def release!
      @window = nil
      @webdialog = nil
      nil
    end

  end # class
end # module